Assets/Scripts/AI/GOAP/GoapPlanner.cs

```diff
@@ -1,17 +1,126 @@
-using System.Collections;
-using System.Collections.Generic;
-using UnityEngine;
-
-namespace ProjetoIA.GOAP
-{
-    public static class GoapPlanner
-    {
-        public static Queue<GoapAction> Plan(GoapAgent agent, 
-                                                List<GoapAction> knowledge, 
-                                                IGoapWorldKnowledge worldKnowledge, 
-                                                GoapGoal goal)
-        {
-            return null;
-        }
-    }
-}
+using System.Collections;
+using System.Collections.Generic;
+using UnityEditor.Experimental.GraphView;
+using UnityEngine;
+
+namespace ProjetoIA.GOAP
+{
+    public class GoapPlanner
+    {
+        private PlanNode[] graph;
+
+        public GoapPlanner(GoapAction[] allActions)
+        {
+            BuildGraph(allActions);
+        }
+
+        // Create all nodes, Calculate all nodes 'solve-by' relations
+        private void BuildGraph(GoapAction[] allActions)
+        {
+            graph = new PlanNode[allActions.Length];
+
+            List<PlanNode> connections = new List<PlanNode>(); // Remember to reset
+            for (int i = 0; i < allActions.Length; i++) graph[i] = new PlanNode(allActions[i]);
+            for (int i = 0; i < allActions.Length; i++) // i is solved by j
+            {
+                for (int j = 0; j < allActions.Length; j++)
+                {
+                    if (i == j) continue;
+                    if (allActions[i].IsValid(allActions[j].GetExpectedEffects())) graph[i].connections.Add(graph[j]);
+                }
+            }
+        }
+
+        public Queue<GoapAction> Plan(IGoapWorldKnowledge worldKnowledge, GoapGoal goal)
+        {
+            List<PlanNode> goalNodes = new List<PlanNode>();
+            foreach (PlanNode node in graph) if (node.Action.GetExpectedEffects() == goal.Objective()) goalNodes.Add(node);
+
+            List<PlanNode> inversePlan = Dijkstra(goalNodes, worldKnowledge);
+
+            Queue<GoapAction> planQueue = new Queue<GoapAction>();
+            for (int i = inversePlan.Count; i > 0; i--) planQueue.Enqueue(inversePlan[i].Action);

```

Comment:
Isso aqui da erro, mudar para:
```suggestion
            for (int i = inversePlan.Count - 1; i >= 0; i--) planQueue.Enqueue(inversePlan[i].Action);
```